### PR TITLE
[FLINK-16202][table] Support JSON_QUERY

### PR DESCRIPTION
**Pull request by Airblader**
## What is the purpose of the change

This introduces support for `JSON_QUERY` in both Table API & SQL.

supersedes #11188

## Verifying this change

This change added tests and can be verified as follows:

* `JsonFunctionsITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs + JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/JsonQueryConverter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        final List<RexNode> operands = new LinkedList<>();
+        operands.add(context.toRexNode(call.getChildren().get(0)));
+        operands.add(context.toRexNode(call.getChildren().get(1)));
+
+        final SqlJsonQueryWrapperBehavior wrappingBehavior =
+                ((ValueLiteralExpression) call.getChildren().get(2))
+                        .getValueAs(JsonQueryWrapper.class)
+                        .map(this::convertWrappingBehavior)
+                        .orElseThrow(() -> new TableException("TODO"));
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/JsonQueryConverter.java:51]: TODO :-)

**Comment by Airblader** [comment]
@flinkbot run azure

File: flink-python/pyflink/table/tests/test_expression.py [hunk 2]
```diff
@@ -207,6 +207,11 @@ def test_expression(self):
                                                   JsonValueOnEmptyOrError.DEFAULT, 42,
                                                   JsonValueOnEmptyOrError.ERROR, None)))
 
+        self.assertEqual("JSON_QUERY('{}', '$.x', WITHOUT, NULL, EMPTY_ARRAY)",
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-python/pyflink/table/tests/test_expression.py:210]: I just noticed that Calcite has also named it like that, that supports my comment 😅

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/BuiltInMethods.scala [hunk 3]
```diff
@@ -24,12 +24,14 @@ import org.apache.flink.table.runtime.functions._
...
 import org.apache.calcite.runtime.{JsonFunctions, SqlFunctions}
-import org.apache.calcite.sql.{SqlJsonExistsErrorBehavior, SqlJsonValueEmptyOrErrorBehavior}
+import org.apache.calcite.sql.{SqlJsonExistsErrorBehavior, SqlJsonQueryEmptyOrErrorBehavior,
+  SqlJsonQueryWrapperBehavior, SqlJsonValueEmptyOrErrorBehavior}
 
 import java.lang.reflect.Method
 import java.lang.{Byte => JByte, Integer => JInteger, Long => JLong, Short => JShort}
 import java.util.TimeZone
 
+
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/BuiltInMethods.scala:34]: undo?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/JsonFunctionsITCase.java [hunk 4]
```diff
@@ -313,4 +321,121 @@ private static TestSpec jsonValue() throws Exception {
                                 true,
                                 DataTypes.BOOLEAN().notNull()));
     }
+
+    private static List<TestSpec> jsonQuery() throws Exception {
+        final InputStream jsonResource =
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/JsonFunctionsITCase.java:326]: nit: move this into a helper method for reading a string, other tests use the same lines too?

**Review by twalthr — COMMENTED** [review]
Thanks @Airblader. I just had a couple of minor comments. I guess we can merge it in the next iteration.

File: docs/data/sql_functions.yml [hunk 5]
```diff
@@ -667,6 +666,46 @@ json:
       JSON_VALUE('{"a": true}', 'strict $.b'
           DEFAULT FALSE ON ERROR)
       ```
+  - sql: JSON_QUERY(jsonValue, path [ { WITHOUT | WITH CONDITIONAL | WITH UNCONDITIONAL } [ ARRAY ] WRAPPER ] [ { NULL | EMPTY ARRAY | EMPTY OBJECT | ERROR } ON EMPTY ] [ { NULL | EMPTY ARRAY | EMPTY OBJECT | ERROR } ON ERROR ])
+    table: STRING.jsonQuery(path [, JsonQueryWrapper [, JsonQueryOnEmptyOrError, JsonQueryOnEmptyOrError ] ])
+    description: |
+      Extracts JSON values from a JSON string.
+
+      The result is always returned as a `STRING`. The `RETURNING` clause is currently not supported.
```
[the remark this answers is not in the document]
Comment (Airblader) [reply docs/data/sql_functions.yml:674]: Yes, Calcite simply doesn't seem to support that.

File: docs/data/sql_functions_zh.yml [hunk 6]
```diff
@@ -799,6 +798,46 @@ json:
...
+
+      ```
+      -- '{ "b": 1 }'
+      JSON_QUERY('{ "a": { "b": 1 } }', '$.a')
+      -- '[1, 2]'
+      JSON_QUERY('[1, 2]', '$')
+
+      -- Wrap result into an array
+      -- '[{}]'
+      JSON_QUERY('{}', '$' WITH CONDITIONAL ARRAY WRAPPER)
```
[the remark this answers is not in the document]
Comment (Airblader) [reply docs/data/sql_functions_zh.yml:823]: The `ARRAY` keyword is optional and only for expresiveness; there is no difference in behavior either way.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
  */
 @PublicEvolving
 public enum JsonQueryWrapper implements TableSymbol {
-    WITHOUT,
+    WITHOUT_ARRAY,
     CONDITIONAL,
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java:32]: `CONDITIONAL_ARRAY` and `UNCONDITIONAL ARRAY` as well?

**Review by twalthr — COMMENTED** [review]
Two last comments.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java (hunk 7 above)
Comment (Airblader) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java:32]: Hm, yeah. Originally I thought it'd be nice to keep the names more concise, but I guess it's better to stay expressive here.

File: docs/data/sql_functions.yml [hunk 8]
```diff
@@ -685,6 +685,8 @@ json:
       JSON_QUERY('{ "a": { "b": 1 } }', '$.a')
       -- '[1, 2]'
       JSON_QUERY('[1, 2]', '$')
+      -- NULL
+      JSON_QUERY((NULL AS STRING), '$')
```
[the remark this answers is not in the document]
Comment (Airblader) [reply docs/data/sql_functions.yml:689]: :see_no_evil:

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java (hunk 7 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/JsonQueryWrapper.java:32]: yes, because it is very specific behavior

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**